### PR TITLE
Container Rehydration from a snapshot taken from detached container.

### PR DESCRIPTION
**Pull request by jatgarg**
Issue: #2692.
Solution:

1.) Loader's createDetachedContainer now takes either iFluidCodeDetails or snapshot. If snapshot is provided, we load from the snapshot.
2.) We keep ripping through the snapshot at container runtime layer and component runtime layer and create local component and channels from the snapshot.
3.) We support delay loading when we load from the snapshot. So when we take the summary for the container to attach, if the component/dds is loaded then we ask them for the summary otherwise we just use the previous summary from which it loaded.
4.) At dds level, we have added a loadLocal api where we load the dds from snapshot but we don't assign it services.
5.) All the components loaded from snapshot are declared as bind but not attached.

### CODE REVIEW

File: packages/runtime/datastore/src/localChannelContext.ts [hunk 1]
```diff
@@ -84,23 +113,95 @@ export class LocalChannelContext implements IChannelContext {
...
-
+        if (this.isLoaded) {
+            assert(this.channel, "Channel should be there if loaded!!");
+            this.channel.connect(this.services);
+        }
         this.attached = true;
     }
+
+    private get services() {
+        if (this._services === undefined) {
```
Comment (jatgarg) [inline packages/runtime/datastore/src/localChannelContext.ts:172]: https://github.com/microsoft/FluidFramework/pull/3061#discussion_r483136761

File: packages/loader/container-loader/src/container.ts [hunk 2]
```diff
@@ -193,7 +202,7 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         options: any,
         scope: IFluidObject,
         loader: Loader,
-        source: IFluidCodeDetails,
+        source: DetachedContainerSource,
         serviceFactory: IDocumentServiceFactory,
```
Comment (jatgarg) [inline packages/loader/container-loader/src/container.ts:206]: https://github.com/microsoft/FluidFramework/pull/3061#discussion_r483120607

**Review by anthony-murphy — CHANGES_REQUESTED** [review]
:clock1:

**Review by anthony-murphy — APPROVED** [review]
:shipit:

**Pull request merged**